### DISCUSSION TITLE
GameCube/Wii/WiiU Support?

### DISCUSSION THREAD

**Discussion by iamhumanipromise** [kth5/archpower#55] [Ideas]
The [Wii U Gitlab](https://gitlab.com/linux-wiiu/linux-wiiu?ssp=1&darkschemeovr=1&setlang=en-US&safesearch=moderate) project page seems detailed, albeit on a 4.x kernel. 

~~The GameCube/Wii [GameCube/Wii Project](https://github.com/neagix/wii-linux-ngx?ssp=1&darkschemeovr=1&setlang=en-US&safesearch=moderate) has a stable 2.6.x kernel but then also experimental/testing branches for 3.x and 4.x thankfully.~~

Would be cool to see these with Arch.

**Comment by kth5** [comment]
There won't be any support for this, sorry.

The Wii with its limited amount of RAM will barely give you a Kernel with an optimized initramfs and a `dash` shell perhaps before you'll end up swapping. The bare minimum for the Arch POWER  installer is 180MiB [1] and the Wii will barely give you 32 free after booting a recent kernel with initramfs.

Custom bootstraps may get the requirement down to within 48MiB I'm sure but it's fairly useless at that point.

Linux 2.6 and and and old Debian isn't such a bad idea on a Wii, bloat killed the video game stars...



[1} https://github.com/kth5/archpower/wiki/Powerpc-%7C-Minimum-Requirements

**Reply by iamhumanipromise** [reply]
That’s Fair. I crossed off the GameCube/Wii part. How about Wii U? It has 3 PPC cores, 2GB RAM, and an ATI GPU.

**Reply by kth5** [reply]
Thats already done but impractical due to the very outdated out of tree kernel (4.19).

If you decompress all firmware and kernel module files, it currently works just fine.

![IMG_1237](https://github.com/kth5/archpower/assets/1784992/04d82efe-67b1-4dd8-8b51-264a6b0f7f59)

2 out of 3 cores are disabled due to CPU hardware bugs btw.

**Reply by Oichkatzelesfrettschen** [reply]
> Thats already done but impractical due to the very outdated out of tree kernel (4.19).
> 
> 
> 
> If you decompress all firmware and kernel module files, it currently works just fine.
> 
> 
> 
> ![IMG_1237](https://github.com/kth5/archpower/assets/1784992/04d82efe-67b1-4dd8-8b51-264a6b0f7f59)
> 
> 
> 
> 2 out of 3 cores are disabled due to CPU hardware bugs btw.

It looks like the Wii Linux project is active [Wii-Linux](https://github.com/Wii-Linux/wii-linux-ngx) and it seems to have a very active [Discord Community](https://discord.gg/MDycbcYM). In case anyone wanted to consider that tiny kernel and then slightly Archify it. Swapping on an SD card would be necessary for more than tiny CLI and TUI or other light-weight apps, but it can still run a system and a UI and other "stuff" under 72MB. 

## My understanding: 

Right now this project is geared toward Void Linux PPC / Void PPC with the VoidPPC750 builds (G3/Wii-era) CPUs being targeted. 

- A big issue that remains is writing an XF86/DRI driver for the graphics... with its weirdly unique architecture it's hard to even get 2D acceleration working unless someone reverse engineers the hardware / Wii driver. 

The project is run by a high school student who has classes starting again soon, so pull requests by devs and active testers will be appreciated as well. 

_Anyway. Linux on the Wii is there. It works. _

----

With this being said, it's Linux 4.5 - he wants a fully functional system before moving to an LTS 4.19, 5.4, 5.10 kernel, etc. 

So: once devs get this cleaned up and tidy, it will be time to move to a modern kernel... and resolve more bugs. 

Anyway. Lots to do. But fun to tinker with, check out and star the project, and maybe use it as inspiration for your Wii U and maybe even GameCube!

**Comment by iamhumanipromise** [comment]
If you need a Wii U let me know, can send: but looks like you have one?

**Comment by hilga007** [comment]
**EDIT: I just noticed the comment above already refers to the Wii Linux project.** Whoops. ***However, This is an update regarding them rebasing to ArchPOWER and some thoughts on the matter.***

Hey kth5,

I'm a huge fan of both ArchPOWER and the Wii Linux project. (It runs on my G4 iBooks and my G5 iMac; with a POWER4 system on the way!)

I recently noticed that the [Wii Linux](https://wii-linux.org/) team is **rebasing their project on ArchPOWER** per their [Discord](https://discord.com/invite/D9EBdRWzv2), and I think this could be an awesome opportunity for both communities to do some collab action and grow together. **OR: Are you both already collaborating?** 

If not yet, here are just a few ideas on how ya'll could make this partnership beneficial for each other, enthusiasts and users:

- It would be great if ArchPOWER could incorporate some of the patches from Wii Linux to better support Wii hardware; perhaps host the Wii kernel? 
- The Wii Linux team is gaining some solid experience with device-specific kernel optimizations that could really boost ArchPOWER’s performance down the line. 
- By working together, both projects could _attract more developers and enthusiasts_ to each, expanding the availability of a *POWER*ful OS across the PowerPC/POWER architecture ecosystem. (Maybe that means more affordable Talos systems someday! xD)
- Once Wii Linux on ArchPOWER is _polished and well-documented_, adding **Wii-specific installation guides** to the ArchPOWER website would be super helpful for users.

Centralizing projects like Wii Linux and PPC Linux 604 within ArchPOWER, along with optimizing for POWER9 and POWER10, can foster innovation and standardization across the POWER ISA. This collaboration can serve as a model for other niche POWER projects to develop tailored kernels and contribute optimizations, thereby enhancing the entire POWER ecosystem.

I'd love to hear your thoughts on this and discuss how we can move forward together. Looking forward to your feedback!

Best regards!

**Comment by techflashYT** [comment]
Edit 1: correct a minor logical mistake I made when rewriting the first comment, I updated my wording from "the patches don't function" to "the patches only function", but forgot to flip the less/greater than sign around to compensate.
Edit 2: Note that it would be possible to get the Wii Linux kernel working, just needs a custom package - it _is_ impossible to do with ArchPOWER's stock kernel.

Hello!  Wii Linux founder and maintainer here.  I've been alerted to this thread by eirikrhinnaskir on Discord, and would like to chime in.

Me and kth5 are _not yet_ collaborating to get ArchPOWER working on the Wii - I'm currently working on adding support via a custom repo all on my own.  However, I _would_ very much be open to collaboration if interested.
Some comments about @hilga007 's ideas:
- Integrating the Wii patches into ArchPOWER's kernel is not possible.  The patches only function properly on kernel versions <= 4.5 yet, and in fact, if built for the Wii, will **not** run on any other device.  Making a custom kernel package (like I'm already doing) would be necessary.
- Not really, I don't have the time, nor experience, to do optimizations like that.  The most I've really done for Wii Linux, is to force PPC750 optimizations in GCC when compiling for a GameCube/Wii.
- That would be neat, again, would gladly collaborate.
- I would gladly write an install guide for the ArchPOWER wiki when my work getting it running is finalized (I'm expecting it to be any day now).

Thanks!

**Reply by kth5** [reply]
@techflashYT If you would like I could throw in a 1 core (4 way SMT on POWER9) `powerpc*` VM with some RAM in for your project if you prefer native building. I have a few helper scripts that manage remote `makerchrootpkg` so working with `PKGBUILD` can feel fairly local as well.

**Reply by techflashYT** [reply]
That would be super useful, I very much appreciate it!  I've been considering grabbing either a G5 Quad or a POWER8 server to do some building on, as it's significantly easier than cross compiling, but haven't been able to come up with the cash.

**Reply by kth5** [reply]
I'm located in `CEST` timezone wise and I have a sess' planned for tonight working on some pretty basic stuff like toolchain updates and bootimage generation. Thus I'll be around and we can see if we can get you set up. 

See you on Discord then? ;)

**Comment by kth5** [comment]
Hey everyone,

it's great news to me that WiiLinux is considering this as a base for their operations, we haven't yet dicussed any collaboration but I'm open to it.

* Regarding the state of things on Wii(U):

I have a rootfs tarball for my WiiU at home but I did not follow it along more since a single core already struggling and having to run `fbdev` along with it was quite a pain. In addition, ArchPOWER (like Archlinux) relies on compressed kernel modules/firmare which in the then current version of linux (4.19) was not yet supported. I also had issues with `zstd` for initramfs and had to use `gzip` instead etc etc. There are many quirks that need touching to get things going properly.

Recently though, I have read (it's on my TODO list) that someone came up with a method to enable all 3 cores without having to rebuild the entire userland to accomodate for the CPU bugs. Which I would say is one showtopper out of the way for me as supporting 4 architectures is more than enough to handle for me already and adding a 5th for just the WiiU sounds overkill.

* Regarding kernels in ArchPOWER

ArchPOWER always had multiple kernels per architecture available:
https://github.com/kth5/archpower/wiki/Available-kernels

Another `linux-wiiu` for instance wouldn't be much of a problem. Then again, the kernel should as best as possible fit in to the rest of the bootchain (firmware, mkinitcpio etc etc).

* Regarding optimization for powerpc 32bit BE

ArchPOWER builds are produced via `-O2 -pipe` which should give fairly decent backwards compatible binaries. However, many projects - examples are `ffmpeg`, `eigen`, `vlc` etc - do bring custom Altivec support. Most of the time it's work to disable those and I have personally did it were meaningful enough to make the effort worthwhile, yet not everywhere.

The idea is that the `base` & `base-devel` package groups should always be able to run on 604 or higher. For the rest it isn't guaranteed as of right now.

**Reply by kth5** [reply]
Additionally, back in the ArchPPC days we had additional "overlay" package repositories by various users.

Much like `[testing]`, those contained packages that would add on or replace those in repositories that were configured below them in `pacman.conf`.

I would personally not go as far as to replace `gcc` or `glibc` and expect the rest of the system to still work fine but it is certainly possible.

**Reply by techflashYT** [reply]
Sadly, as previously mentioned in this discussion, I'm almost entirely positive that following the standard boot stack would be impossible on the OG Wii.  Our current method of booting ("Gumboot") does not offer support for loading an initrd, and recompiling it to add support has not yet been figured out.  It's unkown what compiler was initially used when neagix was developing it, and it fails to compile (or for some, just fails to boot) even with similar versions of GCC.
Even if an initrd _could_ be loaded, the Wii just doesn't have enough RAM.  I maintain a weird sort of hybrid of an initrd and (functionally, at least) GRUB to allow for the user to choose an OS to boot, and know about issues with booting it.  It's a little squashfs image that sits on SD, and is mounted and `switch_root`ed into by an initrd that's baked into the kernel.  However, even just a small buildroot env with a shell and some basic recovery programs + the kernel modules, easily can reach uncompressed sizes balooning to half the Wii's total RAM capacity, and uncompressed sizes reaching up to 20MB over.

However, the current method of booting (Gumboot -> kernel -> pre-baked initrd image stored, and compressed, in the kernel -> wiilinux/${shorthand_of_kernel_ver}.ldr on SD -> boot menu -> distro chosen by the user) _does_ work fairly well, and is how I currently how I boot ArchPOWER, and how I plan to eventually release it, unless some better solution is made.

**Comment by kth5** [comment]
As per announcement the WiiU side of things may have been solved already:
https://github.com/kth5/archpower/discussions/91

**Comment by techflashYT** [comment]
Just figured I would post an update here for anyone else interested;  My unofficial ArchPOWER for the Wii images are out, and by all accounts seem to work great.
An official rootfs and packages, doing things the "right way", are pending some kernel fixes on my end, but should be out _eventually_.  No ETA on that though, no idea how long it'll take to resolve.
GameCube support is on quite an infinite hiatus until I actually get one so I can test the kernel on it.  No plans on that coming to my unofficial images, nor officially to ArchPOWER, any time soon.

**Comment by iamhumanipromise** [comment]
Wonderful news!!

**Comment by techflashYT** [comment]
By the way, new update here.  With the release of Wii-Linux 4.19, the `linux-wii` package on now ArchPOWER works, though is not out yet.  An official release is planned soon, with the following big improvements:
- USB Gecko serial console works again as a proper TTY
- entire EXI stack overhauled to make it less crash-prone, and hotplugging works
- nearly 10k lines of legacy code dropped
- rvl-di (DVD driver) works once again (hopefully can make it in time for release, it doesn't actually work quite yet)
